Test failures

While working a little bit on linkinator on my Windows dev VM I notice an issue if linkinator is already installed globally:

```
  1) cli
       "before all" hook for "should show output for failures":
     Error: Command failed with exit code 1: npm link
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@2.1.3 (node_modules\fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.1.3: wanted {"os":"darwin","arch":"any"} (current: {"os":"win32","arch":"x64"})

npm ERR! code EEXIST
npm ERR! path C:\Users\xmr\AppData\Roaming\npm\node_modules\linkinator\build\src\cli.js
npm ERR! dest C:\Users\xmr\AppData\Roaming\npm\linkinator.cmd
npm ERR! EEXIST: file already exists, cmd shim 'C:\Users\xmr\AppData\Roaming\npm\node_modules\linkinator\build\src\cli.js' -> 'C:\Users\xmr\AppData\Roaming\npm\linkinator.cmd'
npm ERR! File exists: C:\Users\xmr\AppData\Roaming\npm\linkinator.cmd
npm ERR! Remove the existing file and try again, or run npm
npm ERR! with --force to overwrite files recklessly.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\xmr\AppData\Roaming\npm-cache\_logs\2021-01-19T07_04_17_258Z-debug.log

> linkinator@0.0.0 prepare C:\Users\xmr\Desktop\linkinator
> npm run compile


> linkinator@0.0.0 compile C:\Users\xmr\Desktop\linkinator
> tsc -p .

audited 1125 packages in 12.906s

80 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities

      at makeError (node_modules\execa\lib\error.js:59:11)
      at handlePromise (node_modules\execa\index.js:114:26)
      at processTicksAndRejections (internal/process/task_queues.js:93:5)
      at async Context.<anonymous> (build\test\test.cli.js:18:9)
          -> C:\Users\xmr\Desktop\linkinator\test\test.cli.ts:19:5
```

```
C:\Users\xmr\Desktop\linkinator>node -v && npm -v
v14.15.4
6.14.10
```

It seems after running `npm t`, linkinator is being installed globally, so subsequent runs fail.

**EDIT**: the root cause must be this https://github.com/JustinBeckwith/linkinator/blob/071a220712aded6325cab827b2ac9a4deb1a4940/test/test.cli.ts#L18-L20

I'd say it'd be best if this was removed completely. That would also speed tests up. Otherwise, there should be an `unlink` after tests are done.

--- a/test/test.cli.ts
+++ b/test/test.cli.ts
@@ -6,18 +6,8 @@ import * as util from 'util';
 import enableDestroy = require('server-destroy');
 import {LinkResult, LinkState} from '../src/index';
 
-// eslint-disable-next-line prefer-arrow-callback
-describe('cli', function () {
+describe('cli', () => {
   let server: http.Server;
-  this.timeout(60_000);
-
-  if (process.env.LINKINATOR_SKIP_CLI_TESTS) {
-    return;
-  }
-
-  before(async () => {
-    await execa('npm', ['link']);
-  });
 
   afterEach(async () => {
     if (server) {
@@ -26,19 +16,27 @@ describe('cli', function () {
   });
 
   it('should show output for failures', async () => {
-    const res = await execa('linkinator', ['test/fixtures/basic'], {
-      reject: false,
-    });
+    const res = await execa(
+      'node',
+      ['build/src/cli.js', 'test/fixtures/basic'],
+      {
+        reject: false,
+      }
+    );
     assert.match(res.stderr, /ERROR: Detected 1 broken links/);
   });
 
   it('should pass successful markdown scan', async () => {
-    const res = await execa('linkinator', ['test/fixtures/markdown/README.md']);
+    const res = await execa('node', [
+      'build/src/cli.js',
+      'test/fixtures/markdown/README.md',
+    ]);
     assert.match(res.stderr, /Successfully scanned/);
   });
 
   it('should allow multiple paths', async () => {
-    const res = await execa('linkinator', [
+    const res = await execa('node', [
+      'build/src/cli.js',
       'test/fixtures/markdown/unlinked.md',
       'test/fixtures/markdown/README.md',
     ]);
@@ -46,14 +44,15 @@ describe('cli', function () {
   });
 
   it('should show help if no params are provided', async () => {
-    const res = await execa('linkinator', {
+    const res = await execa('node', ['build/src/cli.js'], {
       reject: false,
     });
     assert.match(res.stdout, /\$ linkinator LOCATION \[ --arguments \]/);
   });
 
   it('should flag skipped links', async () => {
-    const res = await execa('linkinator', [
+    const res = await execa('node', [
+      'build/src/cli.js',
       '--verbosity',
       'INFO',
       '--skip',
@@ -64,7 +63,8 @@ describe('cli', function () {
   });
 
   it('should provide CSV if asked nicely', async () => {
-    const res = await execa('linkinator', [
+    const res = await execa('node', [
+      'build/src/cli.js',
       '--format',
       'csv',
       'test/fixtures/markdown/README.md',
@@ -73,7 +73,8 @@ describe('cli', function () {
   });
 
   it('should provide JSON if asked nicely', async () => {
-    const res = await execa('linkinator', [
+    const res = await execa('node', [
+      'build/src/cli.js',
       '--format',
       'json',
       'test/fixtures/markdown/README.md',
@@ -83,7 +84,8 @@ describe('cli', function () {
   });
 
   it('should not show links if --silent', async () => {
-    const res = await execa('linkinator', [
+    const res = await execa('node', [
+      'build/src/cli.js',
       '--silent',
       'test/fixtures/markdown/README.md',
     ]);
@@ -91,7 +93,8 @@ describe('cli', function () {
   });
 
   it('should not show 200 links if verbosity is ERROR with JSON', async () => {
-    const res = await execa('linkinator', [
+    const res = await execa('node', [
+      'build/src/cli.js',
       '--verbosity',
       'ERROR',
       '--format',
@@ -105,7 +108,8 @@ describe('cli', function () {
   });
 
   it('should accept a server-root', async () => {
-    const res = await execa('linkinator', [
+    const res = await execa('node', [
+      'build/src/cli.js',
       '--markdown',
       '--server-root',
       'test/fixtures/markdown',
@@ -115,7 +119,8 @@ describe('cli', function () {
   });
 
   it('should accept globs', async () => {
-    const res = await execa('linkinator', [
+    const res = await execa('node', [
+      'build/src/cli.js',
       'test/fixtures/markdown/*.md',
       'test/fixtures/markdown/**/*.md',
     ]);
@@ -123,16 +128,20 @@ describe('cli', function () {
   });
 
   it('should throw on invalid format', async () => {
-    const res = await execa('linkinator', ['./README.md', '--format', 'LOL'], {
-      reject: false,
-    });
+    const res = await execa(
+      'node',
+      ['build/src/cli.js', './README.md', '--format', 'LOL'],
+      {
+        reject: false,
+      }
+    );
     assert.match(res.stderr, /FORMAT must be/);
   });
 
   it('should throw on invalid verbosity', async () => {
     const res = await execa(
-      'linkinator',
-      ['./README.md', '--VERBOSITY', 'LOL'],
+      'node',
+      ['build/src/cli.js', './README.md', '--VERBOSITY', 'LOL'],
       {
         reject: false,
       }
@@ -142,8 +151,8 @@ describe('cli', function () {
 
   it('should throw when verbosity and silent are flagged', async () => {
     const res = await execa(
-      'linkinator',
-      ['./README.md', '--verbosity', 'DEBUG', '--silent'],
+      'node',
+      ['build/src/cli.js', './README.md', '--verbosity', 'DEBUG', '--silent'],
       {
         reject: false,
       }
@@ -153,8 +162,8 @@ describe('cli', function () {
 
   it('should show no output for verbosity=NONE', async () => {
     const res = await execa(
-      'linkinator',
-      ['test/fixtures/basic', '--verbosity', 'NONE'],
+      'node',
+      ['build/src/cli.js', 'test/fixtures/basic', '--verbosity', 'NONE'],
       {
         reject: false,
       }
@@ -166,8 +175,8 @@ describe('cli', function () {
 
   it('should show callstacks for verbosity=DEBUG', async () => {
     const res = await execa(
-      'linkinator',
-      ['test/fixtures/basic', '--verbosity', 'DEBUG'],
+      'node',
+      ['build/src/cli.js', 'test/fixtures/basic', '--verbosity', 'DEBUG'],
       {
         reject: false,
       }
@@ -177,7 +186,8 @@ describe('cli', function () {
   });
 
   it('should allow passing a config', async () => {
-    const res = await execa('linkinator', [
+    const res = await execa('node', [
+      'build/src/cli.js',
       'test/fixtures/basic',
       '--config',
       'test/fixtures/config/skip-array-config.json',
@@ -207,7 +217,8 @@ describe('cli', function () {
     enableDestroy(server);
     await new Promise<void>(r => server.listen(port, r));
 
-    const res = await execa('linkinator', [
+    const res = await execa('node', [
+      'build/src/cli.js',
       '--retry',
       'test/fixtures/retryCLI',
     ]);

--- a/test/test.cli.ts
+++ b/test/test.cli.ts
@@ -6,8 +6,14 @@ import * as util from 'util';
 import enableDestroy = require('server-destroy');
 import {LinkResult, LinkState} from '../src/index';
 
-describe('cli', () => {
+// eslint-disable-next-line prefer-arrow-callback
+describe('cli', function () {
   let server: http.Server;
+  this.timeout(10_000);
+
+  const pkg = require('../../package.json');
+  const linkinator = pkg.bin.linkinator;
+  const node = 'node';
 
   afterEach(async () => {
     if (server) {
@@ -16,27 +22,23 @@ describe('cli', () => {
   });
 
   it('should show output for failures', async () => {
-    const res = await execa(
-      'node',
-      ['build/src/cli.js', 'test/fixtures/basic'],
-      {
-        reject: false,
-      }
-    );
+    const res = await execa(node, [linkinator, 'test/fixtures/basic'], {
+      reject: false,
+    });
     assert.match(res.stderr, /ERROR: Detected 1 broken links/);
   });
 
   it('should pass successful markdown scan', async () => {
-    const res = await execa('node', [
-      'build/src/cli.js',
+    const res = await execa(node, [
+      linkinator,
       'test/fixtures/markdown/README.md',
     ]);
     assert.match(res.stderr, /Successfully scanned/);
   });
 
   it('should allow multiple paths', async () => {
-    const res = await execa('node', [
-      'build/src/cli.js',
+    const res = await execa(node, [
+      linkinator,
       'test/fixtures/markdown/unlinked.md',
       'test/fixtures/markdown/README.md',
     ]);
@@ -44,15 +46,15 @@ describe('cli', () => {
   });
 
   it('should show help if no params are provided', async () => {
-    const res = await execa('node', ['build/src/cli.js'], {
+    const res = await execa(node, [linkinator], {
       reject: false,
     });
     assert.match(res.stdout, /\$ linkinator LOCATION \[ --arguments \]/);
   });
 
   it('should flag skipped links', async () => {
-    const res = await execa('node', [
-      'build/src/cli.js',
+    const res = await execa(node, [
+      linkinator,
       '--verbosity',
       'INFO',
       '--skip',
@@ -63,8 +65,8 @@ describe('cli', () => {
   });
 
   it('should provide CSV if asked nicely', async () => {
-    const res = await execa('node', [
-      'build/src/cli.js',
+    const res = await execa(node, [
+      linkinator,
       '--format',
       'csv',
       'test/fixtures/markdown/README.md',
@@ -73,8 +75,8 @@ describe('cli', () => {
   });
 
   it('should provide JSON if asked nicely', async () => {
-    const res = await execa('node', [
-      'build/src/cli.js',
+    const res = await execa(node, [
+      linkinator,
       '--format',
       'json',
       'test/fixtures/markdown/README.md',
@@ -84,8 +86,8 @@ describe('cli', () => {
   });
 
   it('should not show links if --silent', async () => {
-    const res = await execa('node', [
-      'build/src/cli.js',
+    const res = await execa(node, [
+      linkinator,
       '--silent',
       'test/fixtures/markdown/README.md',
     ]);
@@ -93,8 +95,8 @@ describe('cli', () => {
   });
 
   it('should not show 200 links if verbosity is ERROR with JSON', async () => {
-    const res = await execa('node', [
-      'build/src/cli.js',
+    const res = await execa(node, [
+      linkinator,
       '--verbosity',
       'ERROR',
       '--format',
@@ -108,8 +110,8 @@ describe('cli', () => {
   });
 
   it('should accept a server-root', async () => {
-    const res = await execa('node', [
-      'build/src/cli.js',
+    const res = await execa(node, [
+      linkinator,
       '--markdown',
       '--server-root',
       'test/fixtures/markdown',
@@ -119,8 +121,8 @@ describe('cli', () => {
   });
 
   it('should accept globs', async () => {
-    const res = await execa('node', [
-      'build/src/cli.js',
+    const res = await execa(node, [
+      linkinator,
       'test/fixtures/markdown/*.md',
       'test/fixtures/markdown/**/*.md',
     ]);
@@ -129,8 +131,8 @@ describe('cli', () => {
 
   it('should throw on invalid format', async () => {
     const res = await execa(
-      'node',
-      ['build/src/cli.js', './README.md', '--format', 'LOL'],
+      node,
+      [linkinator, './README.md', '--format', 'LOL'],
       {
         reject: false,
       }
@@ -140,8 +142,8 @@ describe('cli', () => {
 
   it('should throw on invalid verbosity', async () => {
     const res = await execa(
-      'node',
-      ['build/src/cli.js', './README.md', '--VERBOSITY', 'LOL'],
+      node,
+      [linkinator, './README.md', '--VERBOSITY', 'LOL'],
       {
         reject: false,
       }
@@ -151,8 +153,8 @@ describe('cli', () => {
 
   it('should throw when verbosity and silent are flagged', async () => {
     const res = await execa(
-      'node',
-      ['build/src/cli.js', './README.md', '--verbosity', 'DEBUG', '--silent'],
+      node,
+      [linkinator, './README.md', '--verbosity', 'DEBUG', '--silent'],
       {
         reject: false,
       }
@@ -162,8 +164,8 @@ describe('cli', () => {
 
   it('should show no output for verbosity=NONE', async () => {
     const res = await execa(
-      'node',
-      ['build/src/cli.js', 'test/fixtures/basic', '--verbosity', 'NONE'],
+      node,
+      [linkinator, 'test/fixtures/basic', '--verbosity', 'NONE'],
       {
         reject: false,
       }
@@ -175,8 +177,8 @@ describe('cli', () => {
 
   it('should show callstacks for verbosity=DEBUG', async () => {
     const res = await execa(
-      'node',
-      ['build/src/cli.js', 'test/fixtures/basic', '--verbosity', 'DEBUG'],
+      node,
+      [linkinator, 'test/fixtures/basic', '--verbosity', 'DEBUG'],
       {
         reject: false,
       }
@@ -186,8 +188,8 @@ describe('cli', () => {
   });
 
   it('should allow passing a config', async () => {
-    const res = await execa('node', [
-      'build/src/cli.js',
+    const res = await execa(node, [
+      linkinator,
       'test/fixtures/basic',
       '--config',
       'test/fixtures/config/skip-array-config.json',
@@ -217,8 +219,8 @@ describe('cli', () => {
     enableDestroy(server);
     await new Promise<void>(r => server.listen(port, r));
 
-    const res = await execa('node', [
-      'build/src/cli.js',
+    const res = await execa(node, [
+      linkinator,
       '--retry',
       'test/fixtures/retryCLI',
     ]);

